### PR TITLE
Fix exception reporting due to HTTP request errors.

### PR DESCRIPTION
**Pull request by erikjohnston**
These are business as usual errors, rather than stuff we want to log at error.

### CODE REVIEW

File: synapse/app/generic_worker.py [hunk 1]
```diff
@@ -202,9 +202,14 @@ async def on_POST(self, request, device_id):
...
-                self.main_uri + request.uri.decode("ascii"), body, headers=headers
-            )
+            try:
+                result = await self.http_client.post_json_get_json(
+                    self.main_uri + request.uri.decode("ascii"), body, headers=headers
+                )
+            except HttpResponseException as e:
+                raise e.to_synapse() from e
+            except RequestSendFailed as e:
+                raise SynapseError(502, "Failed to talk to master") from e
```
Comment (richvdh) [inline synapse/app/generic_worker.py:212]: "from e" is implicit when inside an `except` block... maybe it's worth including for clarity?

Comment (erikjohnston) [reply synapse/app/generic_worker.py:212]: I don't believe that is true: https://www.python.org/dev/peps/pep-3134/#explicit-exception-chaining

Comment (richvdh) [reply synapse/app/generic_worker.py:212]: I believe it is: https://www.python.org/dev/peps/pep-3134/#implicit-exception-chaining

Comment (richvdh) [reply synapse/app/generic_worker.py:212]: hrm ISWYM, they seem to do slightly different things. How does the result differ?

Comment (erikjohnston) [reply synapse/app/generic_worker.py:212]: I mean, that is slightly more interesting question. TBH I was mostly going by the fact that we bother using `six.raise_from` elsewhere when wrapping like this.

Comment (erikjohnston) [reply synapse/app/generic_worker.py:212]: I wonder if it gets rid of some of the return deferred exceptions as we explicitly set the `__cause__`?

Comment (erikjohnston) [reply synapse/app/generic_worker.py:212]: Asi in, I think the intention is that the implicit style is for "argh we failed to process this exception and raised another one" while the explicit style is for wrapping exceptions.

Comment (richvdh) [reply synapse/app/generic_worker.py:212]: You might be right. In which case, there are probably lots of other places we should be doing the same.

**Pull request merged**